File: frontend/src-capacitor/android/app/src/main/java/nz/matou/app/SecureStoragePlugin.java
package nz.matou.app;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKey;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Encrypted key/value storage for the WebView — the Android counterpart of
 * Electron's safeStorage-backed secureStorage IPC (issue #71).
 *
 * Values live in EncryptedSharedPreferences ("matou_secure"): keys and values
 * are AES-256 encrypted with a master key held in the Android Keystore, so
 * `shared_prefs/matou_secure.xml` on disk is ciphertext. Nothing secret ever
 * touches WebView localStorage. JS contract (see frontend/src/lib/capacitor.ts):
 *
 *   getItem({key})        -> {value: string | null}
 *   setItem({key, value}) -> void
 *   removeItem({key})     -> void
 */
@CapacitorPlugin(name = "SecureStorage")
public class SecureStoragePlugin extends Plugin {

    private static final String TAG = "SecureStorage";
    private static final String PREFS_FILE = "matou_secure";

    // Keystore + EncryptedSharedPreferences setup does disk and crypto work;
    // keep it (and every call) off the main thread, serialized.
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private SharedPreferences prefs;

    private synchronized SharedPreferences prefs() throws Exception {
        if (prefs == null) {
            Context ctx = getContext();
            MasterKey masterKey = new MasterKey.Builder(ctx)
                    .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                    .build();
            prefs = EncryptedSharedPreferences.create(
                    ctx,
                    PREFS_FILE,
                    masterKey,
                    EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                    EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM);
        }
        return prefs;
    }

    private static String requireKey(PluginCall call) {
        String key = call.getString("key");
        if (key == null || key.isEmpty()) {
            call.reject("SecureStorage: 'key' is required");
            return null;
        }
        return key;
    }

    @PluginMethod
    public void getItem(PluginCall call) {
        String key = requireKey(call);
        if (key == null) return;
        executor.execute(() -> {
            try {
                String value = prefs().getString(key, null);
                JSObject ret = new JSObject();
                // JSObject.put(String, null) would drop the key; be explicit so
                // the JS side always sees a `value` field.
                if (value == null) ret.put("value", JSObject.NULL); else ret.put("value", value);
                call.resolve(ret);
            } catch (Exception e) {
                Log.e(TAG, "getItem failed", e);
                call.reject("SecureStorage: getItem failed: " + e.getMessage(), e);
            }
        });
    }

    @PluginMethod
    public void setItem(PluginCall call) {
        String key = requireKey(call);
        if (key == null) return;
        String value = call.getString("value");
        if (value == null) {
            call.reject("SecureStorage: 'value' is required");
            return;
        }
        executor.execute(() -> {
            try {
                if (!prefs().edit().putString(key, value).commit()) {
                    call.reject("SecureStorage: setItem failed to persist");
                    return;
                }
                call.resolve();
            } catch (Exception e) {
                Log.e(TAG, "setItem failed", e);
                call.reject("SecureStorage: setItem failed: " + e.getMessage(), e);
            }
        });
    }

    @PluginMethod
    public void removeItem(PluginCall call) {
        String key = requireKey(call);
        if (key == null) return;
        executor.execute(() -> {
            try {
                if (!prefs().edit().remove(key).commit()) {
                    call.reject("SecureStorage: removeItem failed to persist");
                    return;
                }
                call.resolve();
            } catch (Exception e) {
                Log.e(TAG, "removeItem failed", e);
                call.reject("SecureStorage: removeItem failed: " + e.getMessage(), e);
            }
        });
    }
}

File: frontend/src-capacitor/ios/App/App/SecureStoragePlugin.swift
import Foundation
import Security
import os.log
import Capacitor

/// Keychain-backed key/value storage for the WebView — the iOS counterpart of
/// Electron's safeStorage-backed secureStorage IPC and Android's
/// EncryptedSharedPreferences plugin (issue #71).
///
/// Items are generic passwords under service `nz.matou.app` (account = key),
/// with `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly`: readable once the
/// device has been unlocked after boot (the backend may run while the screen is
/// locked) and never synced to iCloud Keychain — these are device-bound identity
/// secrets. Nothing secret ever touches WebView localStorage. JS contract (see
/// frontend/src/lib/capacitor.ts), identical to the Java plugin:
///
///   getItem({key})        -> {value: string | null}
///   setItem({key, value}) -> void
///   removeItem({key})     -> void
@objc(SecureStoragePlugin)
public class SecureStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SecureStoragePlugin"
    public let jsName = "SecureStorage"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getItem", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setItem", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeItem", returnType: CAPPluginReturnPromise),
    ]

    private static let log = OSLog(subsystem: "nz.matou.app", category: "SecureStorage")
    private static let service = "nz.matou.app"

    // Keychain calls do IPC to securityd; keep them off the main thread, serialized.
    private let queue = DispatchQueue(label: "nz.matou.app.securestorage", qos: .userInitiated)

    private static func baseQuery(_ key: String) -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
    }

    private static func describe(_ status: OSStatus) -> String {
        if let msg = SecCopyErrorMessageString(status, nil) as String? { return "\(msg) (\(status))" }
        return "OSStatus \(status)"
    }

    private func requireKey(_ call: CAPPluginCall) -> String? {
        guard let key = call.getString("key"), !key.isEmpty else {
            call.reject("SecureStorage: 'key' is required")
            return nil
        }
        return key
    }

    @objc func getItem(_ call: CAPPluginCall) {
        guard let key = requireKey(call) else { return }
        queue.async {
            var query = Self.baseQuery(key)
            query[kSecReturnData as String] = true
            query[kSecMatchLimit as String] = kSecMatchLimitOne
            var item: CFTypeRef?
            let status = SecItemCopyMatching(query as CFDictionary, &item)
            switch status {
            case errSecSuccess:
                guard let data = item as? Data, let value = String(data: data, encoding: .utf8) else {
                    call.reject("SecureStorage: getItem failed: stored value is not UTF-8")
                    return
                }
                call.resolve(["value": value])
            case errSecItemNotFound:
                // Always emit a `value` field so the JS side never sees undefined.
                call.resolve(["value": NSNull()])
            default:
                os_log("getItem failed: %{public}@", log: Self.log, type: .error, Self.describe(status))
                call.reject("SecureStorage: getItem failed: \(Self.describe(status))")
            }
        }
    }

    @objc func setItem(_ call: CAPPluginCall) {
        guard let key = requireKey(call) else { return }
        guard let value = call.getString("value") else {
            call.reject("SecureStorage: 'value' is required")
            return
        }
        queue.async {
            let attrs: [String: Any] = [
                kSecValueData as String: Data(value.utf8),
                kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
            ]
            var query = Self.baseQuery(key)
            var status = SecItemUpdate(query as CFDictionary, attrs as CFDictionary)
            if status == errSecItemNotFound {
                query.merge(attrs) { _, new in new }
                status = SecItemAdd(query as CFDictionary, nil)
            }
            guard status == errSecSuccess else {
                os_log("setItem failed: %{public}@", log: Self.log, type: .error, Self.describe(status))
                call.reject("SecureStorage: setItem failed: \(Self.describe(status))")
                return
            }
            call.resolve()
        }
    }

    @objc func removeItem(_ call: CAPPluginCall) {
        guard let key = requireKey(call) else { return }
        queue.async {
            let status = SecItemDelete(Self.baseQuery(key) as CFDictionary)
            // Removing a missing key succeeds, as on Android.
            guard status == errSecSuccess || status == errSecItemNotFound else {
                os_log("removeItem failed: %{public}@", log: Self.log, type: .error, Self.describe(status))
                call.reject("SecureStorage: removeItem failed: \(Self.describe(status))")
                return
            }
            call.resolve()
        }
    }
}
